pkg/operator/controller/gatewayapi-upgradeable/controller.go

```diff
@@ -0,0 +1,166 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+const (
+	controllerName     = "gatewayapi_upgradeable_controller"
+	GatewayAPIAdminKey = "ack-gateway-api-management"
+	GatewayAPIAdminMsg = "This message was generated due to the presence of the Gateway API CRDs in the cluster. Going forward OCP will fully manage the lifecycle of these CRDs and we strongly advise against any external entity attempting to manage them, as this will be blocked and result in failure. Only the platform itself will support management of these CRDs. Any tampering will cause the cluster to enter a 'Degraded' state. The cluster administrator is responsible for the safety of existing Gateway API implementations. An admin gate will inform the administrator of their responsibilities and prevent upgrades until they acknowledge and accept them. For more information, please refer to the [link to doc here pending]."
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// The New function initializes the controller and sets up the watch for the ConfigMap.
+func New(mgr manager.Manager) (controller.Controller, error) {
+	reconciler := &reconciler{
+		client: mgr.GetClient(),
+		cache:  mgr.GetCache(),
+	}
+
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// A predicate filter to watch for specific changes in the ConfigMap.
+	// Verify that the ConfigMap's name and namespace match the expected values.
+	adminGatePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		cm := o.(*corev1.ConfigMap)
+		return cm.Name == operatorcontroller.AdminGatesConfigMapName().Name && cm.Namespace == operatorcontroller.AdminGatesConfigMapName().Namespace
+	})
+
+	// Mapping the events to the admin gate ConfigMap.
+	toAdminGateConfigMap := func(_ context.Context, _ client.Object) []reconcile.Request {
+		return []reconcile.Request{{
+			NamespacedName: operatorcontroller.AdminGatesConfigMapName(),
+		}}
+	}
+
+	// Defining the CRD predicate.
+	crdPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.(*apiextensionsv1.CustomResourceDefinition).Spec.Group == gatewayapiv1.GroupName
+	})
+
+	// Setting up a watch for CRD events.
+	if err := c.Watch(source.Kind[client.Object](mgr.GetCache(), &apiextensionsv1.CustomResourceDefinition{}, handler.EnqueueRequestsFromMapFunc(toAdminGateConfigMap), crdPredicate)); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(source.Kind[client.Object](mgr.GetCache(), &corev1.ConfigMap{}, &handler.EnqueueRequestForObject{}, adminGatePredicate)); err != nil {
```

Comment:
nit: you could move the definition of the `adminGatePredicate` right above here since this is the only place its used.

```suggestion
	adminGatePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
		return o.GetNamespace() == operatorcontroller.AdminGatesConfigMapName().Namespace &&
			o.GetName() == operatorcontroller.AdminGatesConfigMapName().Name
	})

	if err := c.Watch(source.Kind[client.Object](mgr.GetCache(), &corev1.ConfigMap{}, &handler.EnqueueRequestForObject{}, adminGatePredicate)); err != nil {
```